lib/_http_client.js

```diff
@@ -327,16 +327,22 @@ function socketCloseListener() {
 
   // NOTE: It's important to get parser here, because it could be freed by
   // the `socketOnData`.
-  var parser = socket.parser;
-  if (req.res && req.res.readable) {
-    // Socket closed before we emitted 'end' below.
-    if (!req.res.complete) req.res.emit('aborted');
-    var res = req.res;
-    res.on('end', function() {
+  const parser = socket.parser;
+  const res = req.res;
+  if (res) {
+    if (res.readable) {
+      // Socket closed before we emitted 'end' below.
+      if (!res.complete) {
+        res.emit('aborted');
+      }
+      res.on('end', function() {
+        res.emit('close');
+      });
+      res.push(null);
+    } else {
       res.emit('close');
```

Comment:
I'm not sure when this can happen but shouldn't we also emit a `createHangUpError` here similar to when we haven't yet had a response? i.e. `!res.readable === !res`.